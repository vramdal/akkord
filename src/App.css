.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}

.chord-button-row {
  display: flex;
  width: 100%;
  flex-direction: row;
  justify-self: stretch;
  justify-items: stretch;
}

.chord-button {
  flex: 1 1 3em;
}
