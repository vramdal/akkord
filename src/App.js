import TSXApp from "./TSXApp.tsx";
import React from "react";

export default function() {
  return <TSXApp />;
}
