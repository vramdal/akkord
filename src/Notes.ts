export enum BaseTone {
  C = "C",
  CSharp = "CSharp",
  D = "D",
  DSharp = "DSharp",
  E = "E",
  F = "F",
  FSharp = "FSharp",
  G = "G",
  GSharp = "GSharp",
  A = "A",
  ASharp = "ASharp",
  H = "H",
  DFlat = CSharp,
  EFlat = DSharp,
  GFlat = FSharp,
  AFlat = GSharp,
  HFlat = ASharp
}

export type MIDIOctave = -5 | -4 | -3 | -2 | -1 | 0 | 1 | 2 | 3 | 4 | 6; // https://en.wikipedia.org/wiki/Octave#Notation

export interface Tone {
  baseTone: BaseTone;
  octave: MIDIOctave;
}

export {};
