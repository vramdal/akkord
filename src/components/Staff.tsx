import React from "react";
import {cluster, LINE_HEIGHT} from "./subcomponents/Utils";
import {StaffLine} from "./subcomponents/StaffLine";
import {LedgerLines, Note} from "./subcomponents/Note";
import {PreferredAccidentalsMap, Side} from "./subcomponents/Types";
import {Cursor} from "./subcomponents/Cursor";
import {PreferredAccidentalsContextProvider} from "./subcomponents/Accidentals";


const HollowNoteHeadMask = ({ direction } : {direction: Side }) => (
    <mask id={ `hollowNoteHeadMask${direction}`}
          maskContentUnits="objectBoundingBox"
          maskUnits="objectBoundingBox"
    >
        <rect x="0" y="0" width="1" height="1" fill="white"/>
        <ellipse cx=".5" cy=".5" rx=".25" ry=".4"
                 fill="black"
                 transform={`rotate(${40 * (direction === Side.LEFT ? 1 : -1)}, .5, .5)`}/>
    </mask>

);

const SharpSymbolDefinition = ({lineHeight} : {lineHeight: number}) => <symbol id={"sharp"} width={40} height={lineHeight * 2.5}
                                                                               viewBox={`0 0 10 20`}>
    <g className={"sharp"} stroke="white" strokeWidth={2.5}>
        <line x1={2} x2={2} y1={1} y2={20} strokeWidth={1.5}/>
        <line x1={8} x2={8} y1={0} y2={19} strokeWidth={1.5}/>
        <line x1={0} x2={10} y1={6} y2={3} strokeWidth={3.5}/>
        <line x1={0} x2={10} y1={16} y2={13} strokeWidth={3.5}/>
    </g>
</symbol>;

const FlatSymbolDefinition = ({lineHeight} : {lineHeight: number}) => <symbol id={"flat"} width={40} height={lineHeight * 2.5} viewBox={`5 0 10 20`}>
    <text
        x={0}
        y={20}
        lengthAdjust={"spacingAndGlyphs"}
        textLength={20}
        color={"white"}
        style={{font: "bold 30px sans-serif", fill: "white"}}
        strokeWidth={1.5}>â™­</text>
</symbol>;

interface StaffProps {
    children?: any,
    preferredAccidentals?: PreferredAccidentalsMap
}

export default (props: StaffProps) => {
    const staffWidth = (React.Children.count(props.children) * 150 + 100) || 0;
    return (
        <div className={"staff"} style={{"overflowX": "scroll", "width": "100%"}}>
            <svg
                xmlns="http://www.w3.org/2000/svg"
                width={staffWidth}
                height={"300"}
                className={"staff"}
            >
                <HollowNoteHeadMask direction={Side.LEFT}/>
                <HollowNoteHeadMask direction={Side.RIGHT}/>
                <SharpSymbolDefinition lineHeight={LINE_HEIGHT}/>
                <FlatSymbolDefinition lineHeight={LINE_HEIGHT}/>


                <PreferredAccidentalsContextProvider preferredAccidentals={props.preferredAccidentals || {}}>

                    {[1, 3, 5, 7, 9].map(lineIdx => (
                        <StaffLine key={lineIdx} position={lineIdx} width={staffWidth}/>
                    ))}
                    {React.Children.map(props.children, (child, idx) =>
                        <Cursor x={idx * 100} key={`child-${idx}`}>
                            {child}
                        </Cursor>
                    )}
                </PreferredAccidentalsContextProvider>
            </svg>
        </div>
    );
};

export const _testing = { Note, LedgerLines, cluster };
