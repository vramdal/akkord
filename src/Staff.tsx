import React, {useContext} from "react";
import {addToTone, BaseTone, MIDINote, NoteValues, Tone, getAccidental, Accidental} from "./Notes";

type PositionInStaff = number;


export enum Side {
    LEFT = "LEFT",
    RIGHT = "RIGHT"
}

interface StaffLineProps {
  position: PositionInStaff;
  width: number;
  x: number;
}

function positionInStaffToY(position: PositionInStaff) {
  return position * 12 + 48;
}

const StaffLine = (props: StaffLineProps) => (
    <rect
        y={positionInStaffToY(props.position) + 2}
        x={props.x}
        key={props.position}
        width={props.width}
        height="1"
        className={"staff__line"}
    />
);

StaffLine.defaultProps = {
  width: 200,
  x: 0
};

interface NoteHeadProps {
  x: number;
  positionInStaff: PositionInStaff;
  noteValue: NoteValue;
}

const maskReference = (noteValue: NoteValue) : string | undefined => {
  if (noteValue < 0.5) {
    return undefined;
  } else if (noteValue < 1) {
    return `url(#hollowNoteHeadMask${Side.LEFT})`;
  } else {
    return `url(#hollowNoteHeadMask${Side.RIGHT})`;
  }
};

const NoteHead = ({x, positionInStaff, noteValue}: NoteHeadProps) => {
  const rx = 15;
  const ry = 11;
  const y = positionInStaffToY(positionInStaff) + 3;
  return (
      <ellipse
          cx={x}
          cy={y}
          rx={rx}
          ry={ry}
          className={"note__head"}
          fill={"white"}
          mask={maskReference(noteValue)}
      />
  );
};

interface StemProps {
  sortedNotePositions: Array<PositionInStaff>,
  noteValue: NoteValue;
}

const Stem = ({sortedNotePositions}: StemProps) => {
  const topNotePosition = sortedNotePositions[0];
  const bottomNotePosition = sortedNotePositions[sortedNotePositions.length - 1];
  const stemDirection : "up" | "down" = (bottomNotePosition > 5 && bottomNotePosition < 14) ? "up" : "down";
  const startPosition = stemDirection === "up" ? bottomNotePosition - 8 : bottomNotePosition;
  const endPosition = stemDirection === "down" ? topNotePosition + 8 : topNotePosition;
  const y = positionInStaffToY(startPosition) + 5;
  const length = positionInStaffToY(endPosition) - y;
  const cursor = useContext(CursorContext);
  return (
      <rect
          y={positionInStaffToY(startPosition) + 5}
          x={cursor.x}
          width={4}
          height={length}
          className={"staff__line"}
      />
  );
};

type StemSide = Side;

interface NoteProps {
  toneInfo: ToneInfo;
  isTop: boolean;
  isBottom: boolean;
  stemSide: StemSide;
  noteValue: NoteValue;
  accidental?: Accidental;
}

interface ExtensionLinesProps {
  maxExtent: PositionInStaff;
  x: number;
}

const LedgerLines = (props: ExtensionLinesProps) => {
  let positions: Array<number> = [];
  if (props.maxExtent >= 11) {
    positions = new Array(props.maxExtent - 10)
        .fill(0)
        .map((position, idx) => idx + 11)
        .filter(pos => pos % 2 !== 0);
  } else if (props.maxExtent < 0) {
    positions = new Array(props.maxExtent * -1)
        .fill(0)
        .map((position, idx) => idx * -1 - 1)
        .filter(pos => pos % 2 !== 0);
  }
  return (
      <>
        {positions.map(position => (
            <StaffLine
                key={position}
                position={position}
                x={props.x - 20}
                width={40}
            />
        ))}
      </>
  );
};

interface AccidentalProps {
  staffPosition: PositionInStaff;
  accidental: Accidental
  isTop: boolean;
  isBottom: boolean;
  column: number;
  strKey: string;
}

const Sharp = ({staffPosition, column}: AccidentalProps) => {
  const cursor = useContext(CursorContext);
  const startY = positionInStaffToY(staffPosition - 1);
  const endY = positionInStaffToY(staffPosition + 2);
  const xOffset = (30 * column);
  const length = endY - startY - 3;
  return <>
    <g className={'sharp'} stroke="black" width={4}>
      <rect
          y={startY}
          x={cursor.x - 14 + xOffset}
          height={length}
          width={2}
      />
      <rect
          y={startY - 5}
          x={cursor.x - 7 + xOffset}
          height={length}
          width={2}
      />
      <rect
          y={startY + 9}
          x={cursor.x + xOffset + 5}
          width={5}
          height={17}
          transform={`rotate(80, ${cursor.x + xOffset + 5}, ${startY + 3})`}/>

      />
      <rect
          y={startY + 26}
          x={cursor.x - 2 + xOffset + 5}
          width={5}
          height={17}
          transform={`rotate(80, ${cursor.x + xOffset + 5}, ${startY + 20})`}/>
      />
    </g>
  </>;
};

const Note = ({ toneInfo, isBottom, isTop, noteValue}: NoteProps) => {
  const { staffPosition } = toneInfo;
  const xOffset = 0;// (stemSide === Side.RIGHT ? +16 : -12);
  const cursor = useContext(CursorContext);
  return (
      <>
        {(((isBottom && staffPosition >= 11) || (isTop && staffPosition < 0)) && (
            <LedgerLines maxExtent={staffPosition} x={cursor.x + xOffset} />
        )) ||
        undefined}
        <NoteHead
            key={toneInfo.strKey}
            x={cursor.x + (xOffset)}
            positionInStaff={staffPosition}
            noteValue={noteValue}
        />
      </>
  );
};

Note.defaultProps = {
  isTop: true,
  isBottom: true,
  noteValue: NoteValues.QUARTER
};

type NoteValue = number;

interface ChordProps {
  tones: Array<Tone>;
  noteValue: NoteValue;
}

const mapToneToStaffPosition = (tone: Tone): PositionInStaff => {
  const getBasePosition = () => {
    switch (tone.baseTone) {
      case BaseTone.C:
      case BaseTone.CSharp:
        return 11;
      case BaseTone.D:
      case BaseTone.DSharp:
        return 10;
      case BaseTone.E:
      case BaseTone.EFlat:
        return 9;
      case BaseTone.F:
      case BaseTone.FSharp:
        return 8;
      case BaseTone.G:
      case BaseTone.GFlat:
      case BaseTone.GSharp:
        return 7;
      case BaseTone.A:
      case BaseTone.AFlat:
      case BaseTone.ASharp:
        return 6;
      case BaseTone.H:
      case BaseTone.HFlat:
        return 5;
      default:
        throw new Error("Invalid tone: " + JSON.stringify(tone));
    }
  };
  return getBasePosition() - tone.octave * 7;
};

const toneToMidiNote = (tone: Tone): MIDINote => {
  const baseTones = [
    BaseTone.C,
    BaseTone.CSharp,
    BaseTone.D,
    BaseTone.DSharp,
    BaseTone.E,
    BaseTone.F,
    BaseTone.FSharp,
    BaseTone.G,
    BaseTone.GSharp,
    BaseTone.A,
    BaseTone.ASharp,
    BaseTone.H
  ];

  const getNormalizedBaseTone = (baseTone: BaseTone) => {
    switch (baseTone) {
      case BaseTone.EFlat:
        return BaseTone.DSharp;
      case BaseTone.GFlat:
        return BaseTone.FSharp;
      case BaseTone.AFlat:
        return BaseTone.GSharp;
      case BaseTone.HFlat:
        return BaseTone.ASharp;
      default:
        return baseTone;
    }
  };

  const baseTone = getNormalizedBaseTone(tone.baseTone);
  // noinspection UnnecessaryLocalVariableJS
  const midiNote: MIDINote =
      baseTones.findIndex(currentBaseTone => currentBaseTone === baseTone) +
      12 +
      tone.octave * 12;

  return midiNote;
};

const toneToStrKey = (tone: Tone) => tone.baseTone + tone.octave;

const compareMIDINotes = (MIDINoteA: MIDINote, MIDINoteB: MIDINote) =>
    MIDINoteA.valueOf() - MIDINoteB.valueOf();

//const compareTones = (toneA: Tone, toneB: Tone) => toneToMidiNote(toneA) - toneToMidiNote(toneB);

const compareToneInfos = (
    toneInfoA: { midiNote: MIDINote },
    toneInfoB: { midiNote: MIDINote }
) => compareMIDINotes(toneInfoA.midiNote, toneInfoB.midiNote);

interface ToneInfo extends Tone {
  midiNote: MIDINote;
  strKey: string;
  staffPosition: PositionInStaff;
  accidental: Accidental
}

const CursorContext = React.createContext({x: 50});

const Cursor = (props: {x: number, children: any}) =>
    <CursorContext.Provider value={{x: props.x}}>
      {props.children}
    </CursorContext.Provider>
;

// interface ChordProps {
//   tones: Array<Tone>;
// }

function cluster<Placed extends {staffPosition: PositionInStaff}>(objectsInStaff: Array<Placed>, maxDiffInCluster = 1) {
  const clusters: Array<Array<Placed>> = [];
  let currentCluster: Array<Placed> = [];
  for (let i = 0; i < objectsInStaff.length; i++) {
    const currentTone = objectsInStaff[i];
    const previousTone = i > 0 && objectsInStaff[i - 1];
    if (previousTone && (previousTone.staffPosition - currentTone.staffPosition > maxDiffInCluster)) {
      // End cluster
      clusters.push(currentCluster);
      currentCluster = [];
    }
    currentCluster.push(currentTone);
  }
  clusters.push(currentCluster);
  return clusters;
}

function flattenArray<T>(arrays: Array<Array<T>>): Array<T> {
  const result: Array<T> = [];
  return result.concat.apply([], arrays);
}

const determineStemSide = (note: ToneInfo,
                           positionInCluster: number,
                           cluster: Array<ToneInfo>,
                           stemSideForChord: StemSide
) => {
  if (cluster.length === 1) {
    return stemSideForChord;
  } else {
    return (positionInCluster % 2 === 0 && Side.RIGHT) || Side.LEFT;
  }
};

export const Chord = (props: ChordProps) => {
  const partialToneInfos: Array<ToneInfo> = props.tones.map((tone: Tone) => ({
    ...tone,
    midiNote: toneToMidiNote(tone),
    strKey: toneToStrKey(tone),
    staffPosition: mapToneToStaffPosition(tone),
    accidental: getAccidental(tone),
  })).sort(compareToneInfos);

  const noteClusters: Array<Array<ToneInfo>> = cluster(partialToneInfos, 1);
  const accidentalClusters: Array<Array<ToneInfo>> = cluster(partialToneInfos.filter(toneInfo => toneInfo.accidental), 2);

  const stemSideForChord = partialToneInfos[partialToneInfos.length - 1].staffPosition < 5 ? Side.LEFT : Side.RIGHT;

  const notesProps: Array<NoteProps> = flattenArray<NoteProps>(
      noteClusters.map((cluster, clusterIdx) => {
        return cluster.map(
            (toneInfo, positionInCluster) => ({
              toneInfo: {
                ...toneInfo,
              },
              isTop: clusterIdx === noteClusters.length - 1 && positionInCluster === cluster.length - 1,
              isBottom: clusterIdx === 0 && positionInCluster === 0,
              stemSide: determineStemSide(toneInfo, positionInCluster, cluster, stemSideForChord),
              noteValue: props.noteValue,
              accidental: toneInfo.accidental,
            })
        )
      }));

  const accidentalsProps: Array<AccidentalProps> = flattenArray<AccidentalProps>(
      accidentalClusters.map((cluster, clusterIdx) => {
        return cluster.map(
            (toneInfo : ToneInfo, positionInCluster : number) => ({
              accidental: "sharp",
              isTop: clusterIdx === accidentalClusters.length - 1 && positionInCluster === cluster.length - 1,
              isBottom: clusterIdx === 0 && positionInCluster === 0,
              staffPosition: toneInfo.staffPosition,
              column: 3 - positionInCluster % 3,
              strKey: `${toneInfo.strKey}-accidental`,
            })
        )
      }));

  const staffPositions = notesProps.map(noteProps => noteProps.toneInfo.staffPosition);

  const accidentalsStackWidth = 27;
  const noteHeadsStackWidth = 12;
  const stemWidth = 4;

  const cursor = useContext(CursorContext);

  const accidentalsStartX = cursor.x;
  const numAccidentalColumns = Math.max(0, ...(accidentalsProps.map(accidentalProps => accidentalProps.column + 1)));
  const leftNoteHeadsStartX = accidentalsStartX + (numAccidentalColumns) * accidentalsStackWidth;
  const stemX = (notesProps.find(noteProps => noteProps.stemSide === Side.RIGHT) ? leftNoteHeadsStartX + noteHeadsStackWidth : leftNoteHeadsStartX);
  const rightNodeHeadsStartX = stemX + stemWidth + 12;
  console.log("accidentalsStartX, leftNoteHeadsStartX, stemX, rightNodeHeadsStartX, numAccidentalColumns = ", accidentalsStartX, leftNoteHeadsStartX, stemX, rightNodeHeadsStartX, numAccidentalColumns);

  // TODO: Test accidentals
  return (
      <>
        <Cursor x={accidentalsStartX}>
          { accidentalsProps.map((accidentalProps: AccidentalProps) => (
              accidentalProps.accidental === "sharp"
              && <Sharp {...accidentalProps} key={accidentalProps.strKey}/>
          ))}
        </Cursor>
        <Cursor x={leftNoteHeadsStartX}>
          {notesProps.filter(noteProps => noteProps.stemSide === Side.RIGHT).map((noteProps: NoteProps) => (
              <Note key={noteProps.toneInfo.strKey} {...noteProps} />
          ))}
        </Cursor>
        <Cursor x={stemX}>
          <Stem sortedNotePositions={staffPositions} noteValue={props.noteValue}/>
        </Cursor>
        <Cursor x={rightNodeHeadsStartX}>
          {notesProps.filter(noteProps => noteProps.stemSide === Side.LEFT).map((noteProps: NoteProps) => (
              <Note key={noteProps.toneInfo.strKey} {...noteProps} />
          ))}
        </Cursor>

      </>
  );
};

Chord.defaultProps = {
  noteValue: NoteValues.QUARTER
};

const HollowNoteHeadMask = ({ direction } : {direction: Side }) => (
    <mask id={ `hollowNoteHeadMask${direction}`}
          maskContentUnits="objectBoundingBox"
          maskUnits="objectBoundingBox"
    >
      <rect x="0" y="0" width="1" height="1" fill="white"/>
      <ellipse cx=".5" cy=".5" rx=".25" ry=".4"
               fill="black"
               transform={`rotate(${40 * (direction === Side.LEFT ? 1 : -1)}, .5, .5)`}/>
    </mask>

);


export default (props: {children: any}) => {
  const staffWidth = (React.Children.count(props.children) * 150 + 100) || 0;
  return (
      <div className={"staff"}>
        <svg
            xmlns="http://www.w3.org/2000/svg"
            width={staffWidth}
            height="300"
            className={"staff"}
        >
          <HollowNoteHeadMask direction={Side.LEFT} />
          <HollowNoteHeadMask direction={Side.RIGHT} />

          {[1, 3, 5, 7, 9].map(lineIdx => (
              <StaffLine key={lineIdx} position={lineIdx} width={staffWidth} />
          ))}
          {React.Children.map(props.children,(child, idx) =>
              <Cursor x={50 + idx * 150} key={`child-${idx}`}>
                {child}
              </Cursor>
          )}
        </svg>
      </div>
  );
};

interface ScaleProps {
  startTone: Tone,
  noteValue: NoteValue
}

export const MajorThree = ({startTone, noteValue}: ScaleProps) => {
  const second = addToTone(startTone, 2);
  const third = addToTone(second, 1.5);
  const tones = [startTone, second, third];
  return <Chord tones={tones} noteValue={noteValue}/>
};

export const MinorThree = ({startTone, noteValue}: ScaleProps) => {
  const second = addToTone(startTone, 1.5);
  const third = addToTone(second, 2);
  const tones = [startTone, second, third];
  return <Chord tones={tones} noteValue={noteValue}/>
};

export const _testing = { Note, LedgerLines, cluster };
