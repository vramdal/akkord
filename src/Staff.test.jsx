import React from 'react';
import {render, queries} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import App from './App';
import {Chord} from './Staff.tsx';
import ShallowRenderer from 'react-test-renderer/shallow';
import {BaseTone} from "./Notes";


it('should place an A note on the right place', () => {
    const {getByText, getByTestId} = render( < App />, { queries: {...queries} } ) ;

    expect(getByTestId('x:50,pos:5')).toBeInTheDocument();
})
;

describe('Chord', () => {
    const renderer = new ShallowRenderer();
    it('should render a single Note',() => {
        const toneA0 = {baseTone: BaseTone.A, octave: 0};

        const result = renderer.render( <Chord tones = {[toneA0]}/> );

        expect(result.props.children).toHaveLength(1);
        const note = result.props.children[0];
        expect(note.props).toMatchObject({"hasNeighbor": false, tone: expect.objectContaining(toneA0)});
    });

    it('should render two adjacent nodes as neighbors', () => {
        let toneA0 = {baseTone: BaseTone.A, octave: 0};
        let toneD0 = {baseTone: BaseTone.D, octave: 0};
        let toneH0 = {baseTone: BaseTone.H, octave: 0};
        const tones = [toneA0, toneD0, toneH0];

        const result = renderer.render( <Chord tones = {tones}/> );

        expect(result.props.children).toHaveLength(3);
        const noteD0 = result.props.children[0];
        const noteA0 = result.props.children[1];
        const noteH0 = result.props.children[2];
        expect(noteA0.props).toMatchObject({"hasNeighbor": true, tone: expect.objectContaining(toneA0)});
        expect(noteH0.props).toMatchObject({"hasNeighbor": true, tone: expect.objectContaining(toneH0)});
        expect(noteD0.props).toMatchObject({"hasNeighbor": false, tone: expect.objectContaining(toneD0)});
    });
});

