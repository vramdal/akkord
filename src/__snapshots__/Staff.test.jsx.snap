// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chord should render a half note 1`] = `
Array [
  <ellipse
    className="note__head"
    cx={50}
    cy={123}
    fill="white"
    mask="url(#hollowNoteHeadMaskLEFT)"
    rx={15}
    ry={11}
  />,
  <rect
    className="staff__line"
    height={91}
    width={4}
    x={62}
    y={29}
  />,
]
`;

exports[`Chord should render a single Note 1`] = `
Array [
  <ellipse
    className="note__head"
    cx={50}
    cy={123}
    fill="white"
    rx={15}
    ry={11}
  />,
  <rect
    className="staff__line"
    height={91}
    width={4}
    x={62}
    y={29}
  />,
]
`;

exports[`Chord should render two adjacent nodes as neighbors 1`] = `
Array [
  <ellipse
    className="note__head"
    cx={50}
    cy={171}
    fill="white"
    rx={15}
    ry={11}
  />,
  <ellipse
    className="note__head"
    cx={50}
    cy={123}
    fill="white"
    rx={15}
    ry={11}
  />,
  <rect
    className="staff__line"
    height={151}
    width={4}
    x={62}
    y={113}
  />,
  <ellipse
    className="note__head"
    cx={78}
    cy={111}
    fill="white"
    rx={15}
    ry={11}
  />,
]
`;

exports[`Chord should use the stem side of the top-most node for the entire chord 1`] = `
Array [
  <rect
    className="staff__line"
    height={223}
    width={4}
    x={50}
    y={41}
  />,
  <ellipse
    className="note__head"
    cx={66}
    cy={171}
    fill="white"
    rx={15}
    ry={11}
  />,
  <rect
    className="staff__line"
    height="1"
    width={40}
    x={46}
    y={38}
  />,
  <ellipse
    className="note__head"
    cx={66}
    cy={39}
    fill="white"
    rx={15}
    ry={11}
  />,
]
`;

exports[`Staff should render a half note 1`] = `
Array [
  <ellipse
    className="note__head"
    cx={50}
    cy={123}
    fill="white"
    mask="url(#hollowNoteHeadMaskLEFT)"
    rx={15}
    ry={11}
  />,
  <rect
    className="staff__line"
    height={91}
    width={4}
    x={62}
    y={29}
  />,
]
`;

exports[`Staff should render a single chord 1`] = `
<div
  className="staff"
>
  <svg
    className="staff"
    height="300"
    width={250}
    xmlns="http://www.w3.org/2000/svg"
  >
    <mask
      id="hollowNoteHeadMaskLEFT"
      maskContentUnits="objectBoundingBox"
      maskUnits="objectBoundingBox"
    >
      <rect
        fill="white"
        height="1"
        width="1"
        x="0"
        y="0"
      />
      <ellipse
        cx=".5"
        cy=".5"
        fill="black"
        rx=".25"
        ry=".4"
        transform="rotate(40, .5, .5)"
      />
    </mask>
    <mask
      id="hollowNoteHeadMaskRIGHT"
      maskContentUnits="objectBoundingBox"
      maskUnits="objectBoundingBox"
    >
      <rect
        fill="white"
        height="1"
        width="1"
        x="0"
        y="0"
      />
      <ellipse
        cx=".5"
        cy=".5"
        fill="black"
        rx=".25"
        ry=".4"
        transform="rotate(-40, .5, .5)"
      />
    </mask>
    <rect
      className="staff__line"
      height="1"
      width={250}
      x={0}
      y={62}
    />
    <rect
      className="staff__line"
      height="1"
      width={250}
      x={0}
      y={86}
    />
    <rect
      className="staff__line"
      height="1"
      width={250}
      x={0}
      y={110}
    />
    <rect
      className="staff__line"
      height="1"
      width={250}
      x={0}
      y={134}
    />
    <rect
      className="staff__line"
      height="1"
      width={250}
      x={0}
      y={158}
    />
    <ellipse
      className="note__head"
      cx={50}
      cy={123}
      fill="white"
      rx={15}
      ry={11}
    />
    <rect
      className="staff__line"
      height={91}
      width={4}
      x={62}
      y={29}
    />
  </svg>
</div>
`;

exports[`Staff should render empty staff 1`] = `
<div
  className="staff"
>
  <svg
    className="staff"
    height="300"
    width={100}
    xmlns="http://www.w3.org/2000/svg"
  >
    <mask
      id="hollowNoteHeadMaskLEFT"
      maskContentUnits="objectBoundingBox"
      maskUnits="objectBoundingBox"
    >
      <rect
        fill="white"
        height="1"
        width="1"
        x="0"
        y="0"
      />
      <ellipse
        cx=".5"
        cy=".5"
        fill="black"
        rx=".25"
        ry=".4"
        transform="rotate(40, .5, .5)"
      />
    </mask>
    <mask
      id="hollowNoteHeadMaskRIGHT"
      maskContentUnits="objectBoundingBox"
      maskUnits="objectBoundingBox"
    >
      <rect
        fill="white"
        height="1"
        width="1"
        x="0"
        y="0"
      />
      <ellipse
        cx=".5"
        cy=".5"
        fill="black"
        rx=".25"
        ry=".4"
        transform="rotate(-40, .5, .5)"
      />
    </mask>
    <rect
      className="staff__line"
      height="1"
      width={100}
      x={0}
      y={62}
    />
    <rect
      className="staff__line"
      height="1"
      width={100}
      x={0}
      y={86}
    />
    <rect
      className="staff__line"
      height="1"
      width={100}
      x={0}
      y={110}
    />
    <rect
      className="staff__line"
      height="1"
      width={100}
      x={0}
      y={134}
    />
    <rect
      className="staff__line"
      height="1"
      width={100}
      x={0}
      y={158}
    />
  </svg>
</div>
`;
